Automate certificate installation via the cPanel UAPI

DragoonAethis opened the pull request:
This PR adds an extra feature that's not used by Certbot directly, but allows optional automatic installation of the generated certificate. It uses mostly the same configuration/env vars to call the cPanel `SSL/install_ssl` UAPI with the specified domain and the current live certificate.

The README is updated with the additional config vars (which are optional and used for installation only). To test this, add the new config vars, use `certbot certonly ...` to get a fresh cert and run `./cpanel-dns.py install` - the script will print the server's summary, which usually is something like `['The certificate was successfully installed on the domain “example.com”.']`.

Of course, if you'd prefer not to carry this patch, let me know and I'll close this PR - it's technically not something that Certbot can use directly, just a convenient extra.

DragoonAethis committed:
Automate certificate installation via the cPanel UAPI

alexzorin reviewed (commented):
Thanks!

>Of course, if you'd prefer not to carry this patch, let me know and I'll close this PR - it's technically not something that Certbot can use directly, just a convenient extra.

I think it would be good to merge. It would be neat if this could be integrated into the Certbot call like:

    --deploy-hook "/etc/letsencrypt/cpanel-dns.py install TARGET_DOMAIN"

and we could do away with the `CPANEL_TARGET_DOMAIN` variable, in that case.

DragoonAethis commented:
Oh, I didn't know about deploy hooks - I'll try to get it up and running with that tomorrow.

DragoonAethis committed:
Make the certificate installation flow compatible with --deploy-hook

The script can be now used with certbot certonly/run/renew --deploy-hook
"cpanel-dns.py install example.com" to automatically install certs. If
the domain is not provided on the command line, the script will try to
figure out which CPanel domains to configure with the issued cert from
Certbot's RENEWED_DOMAINS - this is not battle-tested, but should work.

The CPANEL_DNS_INSTALL_TARGET_DOMAIN variable is no longer used.

DragoonAethis committed:
Update README with new install notes

DragoonAethis committed:
Use RENEWED_LINEAGE to pick up the cert location/name automatically

Don't try to magic our way out of RENEWED_DOMAINS - RENEWED_LINEAGE just
points to the cert, so grab its name for the most common case.

DragoonAethis commented:
Alright, tomorrow is today :D I read that Certbot sets `RENEWED_LINEAGE` to the renewed cert's live dir and calls the hook multiple times for multiple generated certs, so it now pulls the domain name from the certificate's path. This can be overridden from the command line, too. This gets rid of all the new install variables.

alexzorin commented on README.md:
````diff
@@ -41,7 +43,11 @@ certbot certonly --manual \
 ```
-5. If this succeeds, so should automatic renewal.
+
+You can optionally add `--deploy-hook "/etc/letsencrypt/cpanel-dns.py install` to install the issued certificate on the same CPanel domain as is the certificate's lineage name (in the example above: `example.com`). You can also use `... install example.com` to specify its name if your CPanel domain is different.
````
```suggestion
You can optionally add `--deploy-hook "/etc/letsencrypt/cpanel-dns.py install"` to install the issued certificate on the same cPanel domain as is the certificate's lineage name (in the example above: `example.com`). You can also use `... install example.com` to specify its name if your cPanel domain is different.
```

alexzorin commented on cpanel-dns.py:
```diff
@@ -131,13 +137,46 @@ def remove_record(domain, txt_value):
+            # Autodetect the domain from the certificate lineage path:
+            domain = cert_live_dir.split('/')[-1]
+
+        print(f"Installing certificate for CPanel domain: {domain}")
```
```suggestion
        print(f"Installing certificate for cPanel domain: {domain}")
```

alexzorin commented on cpanel-dns.py:
```diff
@@ -131,13 +137,46 @@ def remove_record(domain, txt_value):
+            domain = sys.argv[2].strip()
+        else:
+            # Autodetect the domain from the certificate lineage path:
+            domain = cert_live_dir.split('/')[-1]
```
```suggestion
            domain = os.path.basename(os.path.normpath(cert_live_dir))
```

alexzorin reviewed (changes requested):
Thanks, this seems to work well. Just a few minor items in this review.

DragoonAethis committed:
Apply suggestions from code review

Co-authored-by: alexzorin <alex@zor.io>

DragoonAethis commented:
All suggestions merged, thanks!
